subprojects/testing-jvm/src/main/java/org/gradle/api/internal/tasks/testing/detection/TestClasspathFactory.java

```diff
@@ -0,0 +1,170 @@
+/*
+ * Copyright 2023 the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.gradle.api.internal.tasks.testing.detection;
+
+import com.google.common.base.Joiner;
+import com.google.common.collect.ImmutableList;
+import org.gradle.api.internal.classpath.ModuleRegistry;
+import org.gradle.api.internal.tasks.testing.JvmTestExecutionSpec;
+import org.gradle.api.internal.tasks.testing.TestFramework;
+import org.gradle.api.internal.tasks.testing.worker.TestClasspath;
+import org.gradle.api.logging.Logger;
+import org.gradle.api.logging.Logging;
+import org.gradle.internal.classpath.ClassPath;
+import org.gradle.util.internal.CollectionUtils;
+
+import java.io.File;
+import java.net.URL;
+import java.util.List;
+import java.util.function.Function;
+
+/**
+ * Constructs the application and implementation classpaths for a test process,
+ * while also optionally loading test framework implementation dependencies from
+ * the distribution.
+ */
```

Comment:
💭 I think this would benefit from a recap of your comment: https://github.com/gradle/gradle/issues/23995#issuecomment-1449140188, or even just a link to that, since it really helps to explain the why of this.

Reply:
Added an `@see` link